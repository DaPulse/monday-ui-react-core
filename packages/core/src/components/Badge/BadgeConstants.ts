import { DialogPosition } from "../../constants";

/**
 * @deprecated
 */
export enum BadgeType {
  INDICATOR = "indicator",
  COUNTER = "counter"
}

/**
 * @deprecated
 */
export enum BadgeAnchor {
  TOP_START = DialogPosition.TOP_START,
  TOP_END = DialogPosition.TOP_END,
  BOTTOM_START = DialogPosition.BOTTOM_START,
  BOTTOM_END = DialogPosition.BOTTOM_END
}

/**
 * @deprecated
 */
export enum BadgeAlignments {
  RECTANGULAR = "rectangular",
  OUTSIDE = "outside",
  CIRCULAR = "circular"
}
