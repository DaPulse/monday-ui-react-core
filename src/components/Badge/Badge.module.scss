@import "../../styles/themes";

$embeddedPosition: 15%;

.badgeWrapper {
  position: relative;

  .badge {
    z-index: 1;
    position: absolute;
    border: 1px solid transparent;
    border-radius: 16px;
    @include theme-prop(border-color, primary-background-color);

    &.topEnd {
      &.corner {
        top: 0;
        right: 0;
        translate: 50% -50%;
      }

      &.outside {
        top: 0;
        right: 0;
        translate: 100% -50%;
      }

      &.embedded {
        top: $embeddedPosition;
        right: $embeddedPosition;
        translate: 50% -50%;
      }
    }

    &.topStart {
      &.corner {
        top: 0;
        left: 0;
        translate: -50% -50%;
      }

      &.outside {
        top: 0;
        left: 0;
        translate: -100% -50%;
      }

      &.embedded {
        top: $embeddedPosition;
        left: $embeddedPosition;
        translate: -50% -50%;
      }
    }

    &.bottomEnd {
      &.corner {
        bottom: 0;
        right: 0;
        translate: 50% 50%;
      }

      &.outside {
        bottom: 0;
        right: 0;
        translate: 100% -50%;
      }

      &.embedded {
        bottom: $embeddedPosition;
        right: $embeddedPosition;
        translate: 50% 50%;
      }
    }

    &.bottomStart {
      &.corner {
        bottom: 0;
        left: 0;
        translate: -50% 50%;
      }

      &.outside {
        bottom: 0;
        left: 0;
        translate: -100% -50%;
      }

      &.embedded {
        bottom: $embeddedPosition;
        left: $embeddedPosition;
        translate: -50% 50%;
      }
    }
  }
}
