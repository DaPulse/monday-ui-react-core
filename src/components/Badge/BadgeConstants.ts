import { DialogPosition } from "../../constants";

export enum BadgeType {
  INDICATOR = "indicator",
  COUNTER = "counter"
}

export enum BadgeColor {
  PRIMARY = "primary",
  NOTIFICATION = "notification"
}

export enum BadgePosition {
  TOP_START = DialogPosition.TOP_START,
  TOP_END = DialogPosition.TOP_END,
  BOTTOM_START = DialogPosition.BOTTOM_START,
  BOTTOM_END = DialogPosition.BOTTOM_END
}
