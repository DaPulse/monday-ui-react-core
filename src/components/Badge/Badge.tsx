import React, { forwardRef, useRef } from "react";
import cx from "classnames";
import { useMergeRefs } from "../../hooks";
import VibeComponentProps from "../../types/VibeComponentProps";
import VibeComponent from "../../types/VibeComponent";
import { getTestId } from "../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../tests/constants";
import styles from "./Badge.module.scss";
import { Direction } from "./BadgeConstants";
import Indicator, { IndicatorProps } from "../Indicator/Indicator";
import Counter, { CounterProps } from "../Counter/Counter";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import { camelCase } from "lodash-es";

type Position = {
  vertical: Direction.TOP | Direction.BOTTOM;
  horizontal: Direction.LEFT | Direction.RIGHT;
};

type BadgeOptions = typeof Indicator | typeof Counter;

export interface BadgeProps extends VibeComponentProps, React.PropsWithChildren {
  badgeComponent?: BadgeOptions;
  position?: Position;
  gap?: boolean;
  circular?: boolean;
  show?: boolean;
  badgeComponentProps?: CounterProps | IndicatorProps;
}

const Badge: VibeComponent<BadgeProps> & {
  directions?: typeof Direction;
} = forwardRef(
  (
    {
      badgeComponent: BadgeComponent = Indicator,
      position = { vertical: Direction.TOP, horizontal: Direction.RIGHT } as Position,
      gap = false,
      circular = false,
      show = true,
      badgeComponentProps,
      className,
      id,
      "data-testid": dataTestId,
      children
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
    const badgeClassNames = cx(
      styles.badge,
      getStyle(styles, camelCase(`${position.vertical}-${position.horizontal}`)),
      { [styles.gap]: gap, [styles.circular]: circular, [styles.show]: show },
      className
    );

    return (
      <div ref={mergedRef} className={cx(styles.badgeWrapper)} id={id}>
        {children}
        <div className={badgeClassNames} data-testid={dataTestId || getTestId(ComponentDefaultTestId.BADGE, id)}>
          <BadgeComponent {...badgeComponentProps} />
        </div>
      </div>
    );
  }
);

Badge.directions = Direction;

export default Badge;
