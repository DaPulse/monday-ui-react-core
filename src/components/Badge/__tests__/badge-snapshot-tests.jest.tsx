import React from "react";
import renderer from "react-test-renderer";
import Badge from "../Badge";

describe("Badge", () => {
  it("renders with no props with default badge Indicator", () => {
    const tree = renderer.create(<Badge>Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with Counter and counter props", () => {
    const tree = renderer
      .create(
        <Badge type={Badge.types.COUNTER} badgeProps={{ count: 5 }}>
          Child
        </Badge>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders at different position", () => {
    const tree = renderer.create(<Badge position={Badge.positions.BOTTOM_START}>Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with different color", () => {
    const tree = renderer.create(<Badge color={Badge.colors.PRIMARY}>Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders out of horizontal bounds", () => {
    const tree = renderer.create(<Badge outbound>Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with margins from component's bounds", () => {
    const tree = renderer.create(<Badge circular>Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
