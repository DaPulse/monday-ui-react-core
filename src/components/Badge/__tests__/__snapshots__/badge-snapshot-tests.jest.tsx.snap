// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge does not render when show prop is false 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topRight"
    data-testid="badge"
  >
    <div
      className="indicator colorNotification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge renders at different position 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge bottomLeft show"
    data-testid="badge"
  >
    <div
      className="indicator colorNotification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge renders with Counter 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topRight show"
    data-testid="badge"
  >
    <span
      aria-label=" 5"
      aria-labelledby=""
      onMouseDown={[Function]}
    >
      <div
        aria-label="5"
        className="counter sizeLarge kindFill colorPrimary"
      >
        <span
          data-testid="counter"
          id="counter"
        >
          5
        </span>
      </div>
    </span>
  </div>
</div>
`;

exports[`Badge renders with custom Indicator 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topRight show"
    data-testid="badge"
  >
    <div
      className="indicator colorPrimary"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge renders with different children 1`] = `
<div
  className="badgeWrapper"
>
  Other Child
  <div
    className="badge topRight show"
    data-testid="badge"
  >
    <div
      className="indicator colorNotification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge renders with gap and circular 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topRight gap circular show"
    data-testid="badge"
  >
    <div
      className="indicator colorNotification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge renders with no children and props 1`] = `
<div
  className="badgeWrapper"
>
  <div
    className="badge topRight show"
    data-testid="badge"
  >
    <div
      className="indicator colorNotification"
      data-testid="indicator"
    />
  </div>
</div>
`;
